deploy/truenas.sh

```diff
@@ -0,0 +1,186 @@
+  _cert_id=$(echo "$_cert_list" | grep "$_certname" | sed -n 's/.*"\([0-9]\{1,\}\)".*$/\1/p')
+
+  _debug3 _cert_id "$_cert_id"
+
+  _info "Activate Certificate ID: $_cert_id"
+  _activateData="{\"ui_certificate\": \"${_cert_id}\"}"
+  _activate_result="$(_post "$_activateData" "$_api_url/system/general" "" "PUT" "application/json")"
+
+  _debug3 _activate_result "$_activate_result"
+
+  _info "Check if WebDAV certificate is the same as the WEB UI"
+  _webdav_list=$(_get "$_api_url/webdav")
+  _webdav_cert_id=$(echo "$_webdav_list" | grep '"certssl":' | tr -d -- '"certsl: ,')
+
+  if [ "$_webdav_cert_id" = "$_active_cert_id" ]; then
+    _info "Update the WebDAV Certificate"
+    _debug _webdav_cert_id "$_webdav_cert_id"
+    _webdav_data="{\"certssl\": \"${_cert_id}\"}"
+    _activate_webdav_cert="$(_post "$_webdav_data" "$_api_url/webdav" "" "PUT" "application/json")"
+    _webdav_new_cert_id=$(echo "$_activate_webdav_cert" | _json_decode | sed -n 's/.*: \([0-9]\{1,\}\) }$/\1/p')
+    if [ "$_webdav_new_cert_id" -eq "$_cert_id" ]; then
+      _info "WebDAV Certificate update successfully"
+    else
+      _err "Unable to set WebDAV certificate"
+      _debug3 _activate_webdav_cert "$_activate_webdav_cert"
+      _debug3 _webdav_new_cert_id "$_webdav_new_cert_id"
+      return 1
+    fi
+    _debug3 _webdav_new_cert_id "$_webdav_new_cert_id"
+  else
+    _info "WebDAV certificate not set or not the same as Web UI"
+  fi
+
+  _info "Check if FTP certificate is the same as the WEB UI"
+  _ftp_list=$(_get "$_api_url/ftp")
+  _ftp_cert_id=$(echo "$_ftp_list" | grep '"ssltls_certificate":' | tr -d -- '"certislfa:_ ,')
+
+  if [ "$_ftp_cert_id" = "$_active_cert_id" ]; then
+    _info "Update the FTP Certificate"
+    _debug _ftp_cert_id "$_ftp_cert_id"
+    _ftp_data="{\"ssltls_certificate\": \"${_cert_id}\"}"
+    _activate_ftp_cert="$(_post "$_ftp_data" "$_api_url/ftp" "" "PUT" "application/json")"
+    _ftp_new_cert_id=$(echo "$_activate_ftp_cert" | _json_decode | sed -n 's/.*: \([0-9]\{1,\}\) }$/\1/p')
+    if [ "$_ftp_new_cert_id" -eq "$_cert_id" ]; then
+      _info "FTP Certificate update successfully"
+    else
+      _err "Unable to set FTP certificate"
+      _debug3 _activate_ftp_cert "$_activate_ftp_cert"
+      _debug3 _ftp_new_cert_id "$_ftp_new_cert_id"
+      return 1
+    fi
+    _debug3 _activate_ftp_cert "$_activate_ftp_cert"
+  else
+    _info "FTP certificate not set or not the same as Web UI"
+  fi
+
+  _info "Delete old Certificate"
+  _delete_result="$(_post "" "$_api_url/certificate/id/$_active_cert_id" "" "DELETE" "application/json")"
+
+  _debug3 _delete_result "$_delete_result"
+
+  # the command
+  # _restart_UI=$(_get "$_api_url/system/general/ui_restart")
+  # throws the Error 52
+  # for this command direct curl command
+  _info "Reload WebUI from TrueNAS"
+  curl --silent -L --no-keepalive --user-agent "$USER_AGENT" -H "$_H1" "$_api_url/system/general/ui_restart"
```

Comment:
why do we have to use `curl` directly here?
please use `_get()` function instead.

Reply:
as in Line 171-174 described:
there comes the error 52, and I don't know how to catch it.

Reply:
Is the `--no-keepalive` option the reason?

Is there an equivalent for `wget`?

Reply:
I had tested and thought the http sessions would stay open. But curl automatically uses the last session. So the --no-keepalive option is not necessary.

Reply:
But the problem is, that after calling the command `"http://<TruenasIP>/api/v2.0/system/general/ui_restart"`, the Web GUI of the TrueNAS server restarts immediately and does not send any response!

When I call `http://<TruenasIP>/api/v2.0/system/general/ui_restart` with wget it ends up in a loop:
```
HTTP request sent, waiting for response ... No data received.
New try.
HTTP request sent, waiting for response ... No data received.
New try.
...
```

For wget I have to add the option `--tries=1` to prevent the loop
The return code from wget is then 4, error `Network failure.`
The return code from curl is then 52, error `CURLE_GOT_NOTHING (52)`

Then at the end of the deploy output there is one error.
When I edit the script from line 175 to:

```
  _info "Reload WebUI from TrueNAS"
  _restart_UI=$(_get "$_api_url/system/general/ui_restart")
  if [ -n "$_add_cert_result" ] && [ -n "$_activate_result" ]; then
    return 0
  else
    _err "Please refer to https://curl.haxx.se/libcurl/c/libcurl-errors.html for error code: $_ret"
    return 1
  fi
```
As result with curl comes:
```
[Mon Feb 15 23:53:21 CET 2021] Delete old Certificate
[Mon Feb 15 23:53:21 CET 2021] Reload WebUI from TrueNAS
RED LINE: [Mon Feb 15 23:53:22 CET 2021] Please refer to https://curl.haxx.se/libcurl/c/libcurl-errors.html for error code: 52
[Mon Feb 15 23:53:22 CET 2021] Success
```

Ther is a red Line in the Output.

Reply:
could you please show me the log with `curl -vvv`, so that I know what the transfered http headers are.

Reply:
```
[distri@distibutor ~/.acme.sh]$ curl -vvv -H "Authorization: Bearer DEPLOY_TRUENAS_APIKEY" -H "Content-Type: application/json" http://truenas.mydomain.de/api/v2.0/system/general/ui_restart
*   Trying 192.168.1.3:80...
* Connected to truenas.mydomain.de (192.168.1.3) port 80 (#0)
> GET /api/v2.0/system/general/ui_restart HTTP/1.1
> Host: truenas.mydomain.de
> User-Agent: curl/7.74.0
> Accept: */*
> Authorization: Bearer DEPLOY_TRUENAS_APIKEY
> Content-Type: application/json
>
* Empty reply from server
* Connection #0 to host truenas.mydomain.de left intact
curl: (52) Empty reply from server
[distri@distibutor ~/.acme.sh]$
```
I have replaced the real API-Key with DEPLOY_TRUENAS_APIKEY

Reply:
same for wget, but with the option -d and --tries=3 (the default of wget is --tries=20):

```
[distri@distibutor ~/.acme.sh]$ wget -d --tries=3 --header "Authorization: Bearer DEPLOY_TRUENAS_APIKEY" --header "Content-Type: application/json" -O - http://truenas.my-domain.de/api/v2.0/system/general/ui_restart
Setting --tries (tries) to 3
Setting --tries (tries) to 3
Setting --header (header) to Authorization: Bearer DEPLOY_TRUENAS_APIKEY
Setting --header (header) to Authorization: Bearer DEPLOY_TRUENAS_APIKEY
Setting --header (header) to Content-Type: application/json
Setting --header (header) to Content-Type: application/json
Setting --output-document (outputdocument) to -
Setting --output-document (outputdocument) to -
DEBUG output created by Wget 1.20.3 on freebsd12.2.

Reading HSTS entries from /home/distri/.wget-hsts
URI encoding = 'US-ASCII'
iconv UTF-8 -> US-ASCII
iconv outlen=120 inlen=60
converted 'http://truenas.my-domain.de/api/v2.0/system/general/ui_restart' (US-ASCII) -> 'http://truenas.my-domain.de/api/v2.0/system/general/ui_restart' (UTF-8)
--2021-02-17 10:43:01--  http://truenas.my-domain.de/api/v2.0/system/general/ui_restart
Resolving truenas.my-domain.de (truenas.my-domain.de)... 192.168.1.3
Caching truenas.my-domain.de => 192.168.1.3
Connecting to truenas.my-domain.de (truenas.my-domain.de)|192.168.1.3|:80... connected.
Created socket 3.
Releasing 0x0000000800cd31a0 (new refcount 1).

---request begin---
GET /api/v2.0/system/general/ui_restart HTTP/1.1
User-Agent: Wget/1.20.3 (freebsd12.2)
Accept: */*
Accept-Encoding: identity
Host: truenas.my-domain.de
Connection: Keep-Alive
Authorization: Bearer DEPLOY_TRUENAS_APIKEY
Content-Type: application/json

---request end---
HTTP request sent, awaiting response... No data received.
Closed fd 3
Retrying.

--2021-02-17 10:43:03--  (try: 2)  http://truenas.my-domain.de/api/v2.0/system/general/ui_restart
Found truenas.my-domain.de in host_name_addresses_map (0x800cd31a0)
Connecting to truenas.my-domain.de (truenas.my-domain.de)|192.168.1.3|:80... connected.
Created socket 3.
Releasing 0x0000000800cd31a0 (new refcount 1).

---request begin---
GET /api/v2.0/system/general/ui_restart HTTP/1.1
User-Agent: Wget/1.20.3 (freebsd12.2)
Accept: */*
Accept-Encoding: identity
Host: truenas.my-domain.de
Connection: Keep-Alive
Authorization: Bearer DEPLOY_TRUENAS_APIKEY
Content-Type: application/json

---request end---
HTTP request sent, awaiting response... No data received.
Closed fd 3
Retrying.

--2021-02-17 10:43:06--  (try: 3)  http://truenas.my-domain.de/api/v2.0/system/general/ui_restart
Found truenas.my-domain.de in host_name_addresses_map (0x800cd31a0)
Connecting to truenas.my-domain.de (truenas.my-domain.de)|192.168.1.3|:80... connected.
Created socket 3.
Releasing 0x0000000800cd31a0 (new refcount 1).

---request begin---
GET /api/v2.0/system/general/ui_restart HTTP/1.1
User-Agent: Wget/1.20.3 (freebsd12.2)
Accept: */*
Accept-Encoding: identity
Host: truenas.my-domain.de
Connection: Keep-Alive
Authorization: Bearer DEPLOY_TRUENAS_APIKEY
Content-Type: application/json

---request end---
HTTP request sent, awaiting response... No data received.
Closed fd 3
Giving up.

[distri@distibutor ~/.acme.sh]$
```

Reply:
it seems that the server end doesn't repond any thing, even without a `200 OK` status line?

Could it be a bug of the server  end ?

Can you please try to POST an empty string or a random string to the server, to see how it will respond?

Thanks.

Reply:
That is right. There is no respond to this api call `/system/general/ui_restart`.
All other apicalls for example `/system/general` from line 83 are responding with a good respond.
Same in the Lines 109, 116, 121 and so on.

I have opend a bug report at ixSystem, [see here](https://jira.ixsystems.com/browse/NAS-109435?orderby=created+DESC%2C+priority+DESC%2C+updated+DESC).

I could use the [code from here](https://github.com/acmesh-official/acme.sh/pull/3396#discussion_r576462932) and add a extra error text line that this error 52 with curl is "normal" Perhaps a extra link to the bug report. But if some one uses wget for this scipt thats a problem, then there must set the option --tries=1 to wget. Otherwise the restart of the UI would be about 20 times.